Crash report 3.9-rc-4: RejectedExecutionException in MediaUtils.fadeMediaItemImageIntoView

Too many AsyncTask queued (from `MediaUtils.fadeMediaItemImageIntoView`)

```
java.util.concurrent.RejectedExecutionException: Task android.os.AsyncTask$3@364d4662 rejected from java.util.concurrent.ThreadPoolExecutor@3c2503f3[Running, pool size = 9, active threads = 9, queued tasks = 128, completed tasks = 556]
       at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2011)
       at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:793)
       at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1339)
       at android.os.AsyncTask.executeOnExecutor(AsyncTask.java:590)
       at org.wordpress.mediapicker.MediaUtils.fadeMediaItemImageIntoView(MediaUtils.java:123)
       at org.wordpress.mediapicker.source.MediaSourceDeviceImages.getView(MediaSourceDeviceImages.java:143)
       at org.wordpress.mediapicker.MediaSourceAdapter.getView(MediaSourceAdapter.java:83)
       at android.widget.AbsListView.obtainView(AbsListView.java:2344)
       at android.widget.GridView.makeAndAddView(GridView.java:1433)
       at android.widget.GridView.makeRow(GridView.java:361)
       at android.widget.GridView.fillDown(GridView.java:302)
       at android.widget.GridView.fillGap(GridView.java:262)
       at android.widget.AbsListView.trackMotionScroll(AbsListView.java:4972)
       at android.widget.AbsListView$FlingRunnable.run(AbsListView.java:4516)
       at android.view.Choreographer$CallbackRecord.run(Choreographer.java:767)
       at android.view.Choreographer.doCallbacks(Choreographer.java:580)
       at android.view.Choreographer.doFrame(Choreographer.java:549)
       at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:753)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:135)
       at android.app.ActivityThread.main(ActivityThread.java:5336)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
```

Same kind of issue:
```
java.lang.RuntimeException: Unable to start activity ComponentInfo{org.wordpress.android/org.wordpress.android.ui.media.MediaPickerActivity}: java.util.concurrent.RejectedExecutionException: Task android.os.AsyncTask$3@445f4760 rejected from java.util.concurrent.ThreadPoolExecutor@41c95560[Running, pool size = 9, active threads = 9, queued tasks = 128, completed tasks = 386]
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2338)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2390)
       at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:3930)
       at android.app.ActivityThread.access$900(ActivityThread.java:151)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1327)
       at android.os.Handler.dispatchMessage(Handler.java:110)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:5292)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:824)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:640)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by: java.util.concurrent.RejectedExecutionException: Task android.os.AsyncTask$3@445f4760 rejected from java.util.concurrent.ThreadPoolExecutor@41c95560[Running, pool size = 9, active threads = 9, queued tasks = 128, completed tasks = 386]
       at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2011)
       at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:793)
       at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1339)
       at android.os.AsyncTask.executeOnExecutor(AsyncTask.java:590)
       at org.wordpress.android.ui.media.MediaSourceWPImages.removeDeletedEntries(MediaSourceWPImages.java:212)
       at org.wordpress.android.ui.media.MediaSourceWPImages.addWordPressImagesFromCursor(MediaSourceWPImages.java:162)
       at org.wordpress.android.ui.media.MediaSourceWPImages.gather(MediaSourceWPImages.java:49)
       at org.wordpress.mediapicker.MediaSourceAdapter.gatherFromSources(MediaSourceAdapter.java:46)
       at org.wordpress.mediapicker.MediaPickerFragment.generateAdapter(MediaPickerFragment.java:536)
       at org.wordpress.mediapicker.MediaPickerFragment.setMediaSources(MediaPickerFragment.java:338)
       at org.wordpress.mediapicker.MediaPickerFragment.restoreFromBundle(MediaPickerFragment.java:468)
       at org.wordpress.mediapicker.MediaPickerFragment.onCreate(MediaPickerFragment.java:133)
       at android.app.Fragment.performCreate(Fragment.java:1678)
       at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:859)
       at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1062)
       at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1044)
       at android.app.FragmentManagerImpl.dispatchCreate(FragmentManager.java:1848)
       at android.app.Activity.onCreate(Activity.java:902)
       at android.support.v4.app.FragmentActivity.onCreate(FragmentActivity.java:255)
       at android.support.v7.app.ActionBarActivity.onCreate(ActionBarActivity.java:122)
       at org.wordpress.android.ui.media.MediaPickerActivity.onCreate(MediaPickerActivity.java:98)
       at android.app.Activity.performCreate(Activity.java:5264)
       at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1088)
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2302)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2390)
       at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:3930)
       at android.app.ActivityThread.access$900(ActivityThread.java:151)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1327)
       at android.os.Handler.dispatchMessage(Handler.java:110)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:5292)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:824)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:640)
       at dalvik.system.NativeStart.main(NativeStart.java)
```

--- a/WordPress/build.gradle
+++ b/WordPress/build.gradle
@@ -14,7 +14,6 @@ repositories {
     mavenCentral()
     maven { url 'http://wordpress-mobile.github.io/WordPress-Android' }
     maven { url 'http://download.crashlytics.com/maven' }
-    maven { url 'https://oss.sonatype.org/content/repositories/staging' }
 }
 
 apply plugin: 'com.android.application'

--- a/WordPress/build.gradle
+++ b/WordPress/build.gradle
@@ -14,6 +14,7 @@ repositories {
     mavenCentral()
     maven { url 'http://wordpress-mobile.github.io/WordPress-Android' }
     maven { url 'http://download.crashlytics.com/maven' }
+    maven { url 'https://oss.sonatype.org/content/repositories/staging' }
 }
 
 apply plugin: 'com.android.application'
@@ -66,7 +67,7 @@ dependencies {
     compile 'com.crashlytics.android:crashlytics:+'
 
     // Provided by maven central
-    compile ('org.wordpress:mediapicker:1.2.0') {
+    compile ('org.wordpress:mediapicker:1.2.2') {
         exclude group:'com.android.support'
     }
     compile 'com.google.code.gson:gson:2.2.2'

--- a/WordPress/src/main/java/org/wordpress/android/ui/media/MediaSourceWPImages.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/media/MediaSourceWPImages.java
@@ -221,7 +221,7 @@ public class MediaSourceWPImages implements MediaSource {
                 }
 
                 @Override
-                public void performExecution(Object params) {
+                public void startExecution(Object params) {
                     if (!(params instanceof MediaItem)) {
                         throw new IllegalArgumentException("Params must be of type MediaItem");
                     }

--- a/WordPress/src/main/java/org/wordpress/android/ui/media/MediaSourceWPImages.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/media/MediaSourceWPImages.java
@@ -3,6 +3,7 @@ package org.wordpress.android.ui.media;
 import android.content.Context;
 import android.database.Cursor;
 import android.graphics.Bitmap;
+import android.graphics.drawable.Drawable;
 import android.net.Uri;
 import android.os.AsyncTask;
 import android.os.Parcel;
@@ -94,7 +95,7 @@ public class MediaSourceWPImages implements MediaSource {
                     }
 
                     if (imageBitmap == null) {
-                        imageView.setImageResource(R.color.grey_darken_10);
+                        imageView.setImageDrawable(placeholderDrawable(convertView.getContext()));
                         MediaUtils.BackgroundDownloadWebImage bgDownload = new MediaUtils.BackgroundDownloadWebImage(imageView);
                         imageView.setTag(bgDownload);
                         bgDownload.executeOnExecutor(AsyncTask.SERIAL_EXECUTOR, mediaItem.getPreviewSource());
@@ -114,6 +115,14 @@ public class MediaSourceWPImages implements MediaSource {
     @Override
     public boolean onMediaItemSelected(MediaItem mediaItem, boolean selected) {
         return !selected;
+    }
+
+    private Drawable placeholderDrawable(Context context) {
+        if (context != null && context.getResources() != null) {
+            return context.getResources().getDrawable(R.drawable.media_item_placeholder);
+        }
+
+        return null;
     }
 
     private void addWordPressImagesFromCursor(Cursor cursor) {

--- a/WordPress/src/main/java/org/wordpress/android/ui/media/MediaSourceWPImages.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/media/MediaSourceWPImages.java
@@ -21,6 +21,7 @@ import org.wordpress.android.WordPressDB;
 import org.wordpress.android.models.Blog;
 import org.wordpress.mediapicker.MediaItem;
 import org.wordpress.mediapicker.source.MediaSource;
+import org.wordpress.mediapicker.MediaUtils.LimitedBackgroundOperation;
 
 import java.io.IOException;
 import java.net.HttpURLConnection;
@@ -180,11 +181,12 @@ public class MediaSourceWPImages implements MediaSource {
         for (MediaItem mediaItem : existingItems) {
             final boolean callLoaded = existingItems.indexOf(mediaItem) == existingItems.size() - 1;
 
-            AsyncTask<MediaItem, Void, MediaItem> backgroundCheck = new AsyncTask<MediaItem, Void, MediaItem>() {
+            LimitedBackgroundOperation<MediaItem, Void, MediaItem> backgroundCheck =
+                    new LimitedBackgroundOperation<MediaItem, Void, MediaItem>() {
                 int responseCode;
 
                 @Override
-                protected MediaItem doInBackground(MediaItem[] params) {
+                protected MediaItem performBackgroundOperation(MediaItem[] params) {
                     MediaItem mediaItem = params[0];
                     try {
                         URL mediaUrl = new URL(mediaItem.getSource().toString());
@@ -203,7 +205,7 @@ public class MediaSourceWPImages implements MediaSource {
                 }
 
                 @Override
-                public void onPostExecute(MediaItem result) {
+                public void performPostExecute(MediaItem result) {
                     if (mListener != null && result != null) {
                         List<MediaItem> resultList = new ArrayList<>();
                         resultList.add(result);
@@ -217,8 +219,16 @@ public class MediaSourceWPImages implements MediaSource {
                         }
                     }
                 }
+
+                @Override
+                public void performExecution(Object params) {
+                    if (!(params instanceof MediaItem)) {
+                        throw new IllegalArgumentException("Params must be of type MediaItem");
+                    }
+                    executeOnExecutor(THREAD_POOL_EXECUTOR, (MediaItem) params);
+                }
             };
-            backgroundCheck.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, mediaItem);
+            backgroundCheck.executeWithLimit(mediaItem);
         }
     }
 

--- a/WordPress/src/main/java/org/wordpress/android/ui/media/MediaSourceWPVideos.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/media/MediaSourceWPVideos.java
@@ -3,6 +3,7 @@ package org.wordpress.android.ui.media;
 import android.content.Context;
 import android.database.Cursor;
 import android.graphics.Bitmap;
+import android.graphics.drawable.Drawable;
 import android.net.Uri;
 import android.os.Parcel;
 import android.view.LayoutInflater;
@@ -89,7 +90,7 @@ public class MediaSourceWPVideos implements MediaSource {
 
                     imageView.setScaleType(ImageView.ScaleType.CENTER_CROP);
                     if (imageBitmap == null) {
-                        imageView.setImageResource(R.color.grey_darken_10);
+                        imageView.setImageDrawable(placeholderDrawable(convertView.getContext()));
                         MediaUtils.BackgroundDownloadWebImage bgDownload =
                                 new MediaUtils.BackgroundDownloadWebImage(imageView);
                         imageView.setTag(bgDownload);
@@ -111,6 +112,14 @@ public class MediaSourceWPVideos implements MediaSource {
     @Override
     public boolean onMediaItemSelected(MediaItem mediaItem, boolean selected) {
         return !selected;
+    }
+
+    private Drawable placeholderDrawable(Context context) {
+        if (context != null && context.getResources() != null) {
+            return context.getResources().getDrawable(R.drawable.media_item_placeholder);
+        }
+
+        return null;
     }
 
     /**

--- a/WordPress/src/main/res/drawable-xxhdpi/media_item_placeholder.xml
+++ b/WordPress/src/main/res/drawable-xxhdpi/media_item_placeholder.xml
@@ -4,6 +4,6 @@
     xmlns:android="http://schemas.android.com/apk/res/android">
 
     <item
-        android:drawable="@drawable/media_image_placeholder" />
+        android:drawable="@color/grey_darken_10" />
 
 </selector>

--- a/WordPress/src/main/res/drawable-xxhdpi/media_item_placeholder.xml
+++ b/WordPress/src/main/res/drawable-xxhdpi/media_item_placeholder.xml
@@ -1,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<selector
+    xmlns:android="http://schemas.android.com/apk/res/android">
+
+    <item
+        android:drawable="@drawable/media_image_placeholder" />
+
+</selector>

